sdk/sdk.go

```diff
@@ -0,0 +1,162 @@
+// Package tfsdklog provides helper functions for logging from SDKs and
+// frameworks for building Terraform plugins.
+//
+// Plugin authors shouldn't need to use this package; it is meant for authors
+// of the frameworks and SDKs for plugins. Plugin authors should use the tflog
+// package.
+//
+// This package provides very similar functionality to tflog, except it uses a
+// separate namespace for its logs.
+package tfsdklog
+
+import (
+	"context"
+	"io"
+	"os"
+
+	"github.com/hashicorp/go-hclog"
+)
+
+type loggerKey string
+
+var (
+	rootLoggerKey loggerKey = "sdk"
+	stderr        io.Writer
+)
+
+func init() {
+	// When go-plugin.Serve is called, it overwrites our os.Stderr with a
+	// gRPC stream which Terraform ignores. This tends to be before our
+	// loggers get set up, as go-plugin has no way to pass in a base
+	// context, and our loggers are passed around via contexts. This leaves
+	// our loggers writing to an output that is never read by anything,
+	// meaning the logs get blackholed. This isn't ideal, for log output,
+	// so this is our workaround: we copy stderr on init, before Serve can
+	// be called, and offer an option to write to that instead of the
+	// os.Stderr available at runtime.
+	//
+	// Ideally, this is a short-term fix until Terraform starts reading
+	// from go-plugin's gRPC-streamed stderr channel, but for the moment it
+	// works.
+	stderr = os.Stderr
+}
+
+func getRootLogger(ctx context.Context) hclog.Logger {
+	logger := ctx.Value(rootLoggerKey)
+	if logger == nil {
+		return nil
+	}
+	return logger.(hclog.Logger)
+}
+
+func setRootLogger(ctx context.Context, logger hclog.Logger) context.Context {
+	return context.WithValue(ctx, rootLoggerKey, logger)
+}
+
+// New returns a new context.Context that contains a logger configured with the
+// passed options.
```

Comment:
It's a new context, based on the passed context; the original context is not modified. Just as this function returns a new string, it doesn't modify the string:

```go
func hello(name string) string {
  return "hello, "+name
}
```

The distinction is important because it means the return parameter contains the logger, but the argument still doesn't.

Reply:
Ah, TIL [`(context.Context).WithValue()`](https://golang.org/pkg/context/#WithValue) returns a copy.